lib/ansible/module_utils/facts/virtual/linux.py

```diff
@@ -109,6 +109,11 @@ def get_virtual_facts(self):
                 virtual_facts['virtualization_type'] = systemd_container
                 virtual_facts['virtualization_role'] = 'guest'
                 found_virt = True
+        elif os.path.exists('/.dockerenv') or os.path.exists('/.dockerinit'):
+            guest_tech.add('docker')
+            virtual_facts['virtualization_type'] = 'docker'
+            virtual_facts['virtualization_role'] = 'guest'
+            found_virt = True
```

Comment:
```suggestion

        # If docker/containerd has a custom cgroup parent, checking /proc/1/cgroup (above) might fail.
        # https://docs.docker.com/engine/reference/commandline/dockerd/#default-cgroup-parent
        # Fallback to more rudimentary checks.
        if os.path.exists('/.dockerenv') or os.path.exists('/.dockerinit'):
            guest_tech.add('docker')
            if not found_virt:
                virtual_facts['virtualization_type'] = 'docker'
                virtual_facts['virtualization_role'] = 'guest'
                found_virt = True
```

What does this check have to do with systemd? I think it should be separate.
Also, only update `virtualization_type` if `found_virt` is False (similar to other `if not found_virt` blocks in this file).

Reply:
I agree and I have a feeling this fix is going to become more and more important with current and future Linux distributions. I made a pull request as well before I realized this one existed.

https://github.com/ansible/ansible/pull/75340/files#diff-48cedf55779922152790325bc5bccee4ca72603a3866e81345d47352cfcecc32R71-R78